Interop/StubGenerator/src/main/kotlin/org/jetbrains/kotlin/native/interop/gen/StubIrType.kt

```diff
@@ -0,0 +1,181 @@
+    is KotlinClassifierType -> this.toStubIrType()
+    else -> error("Unexpected KotlinType: $this")
+}
+
+private fun KotlinFunctionType.toStubIrType(): StubType =
+        FunctionalType(parameterTypes.map { it.toStubIrType() }, returnType.toStubIrType(), nullable)
+
+private fun KotlinClassifierType.toStubIrType(): StubType {
+    PredefinedTypesHandler.tryExpandPredefinedTypealias(classifier, nullable)?.let { return it }
+    val typeArguments = arguments.map { it.toStubIrType() }
+    return if (expandedType != null) {
+        val expandedType = expandedType.toStubIrType()
+        val name = ClassifierStubType(classifier, typeArguments)
+        AbbreviationStubType(expandedType, name, nullable)
+    } else {
+        ClassifierStubType(classifier, typeArguments, nullable)
+    }
+}
+
+private fun KotlinTypeArgument.toStubIrType(): TypeArgument = when (this) {
+    is KotlinType -> TypeArgumentStub(this.toStubIrType())
+    StarProjection -> TypeArgument.StarProjection
+    else -> error("Unexpected KotlinTypeArgument: $this")
+}
+
+/**
+ * Types that come from kotlinx.cinterop require special handling because we
+ * don't have explicit information about their structure.
+ * For example, to be able to produce metadata-based interop library we need to know
+ * that ByteVar is a typealias to ByteVarOf<Byte>.
+ */
+private object PredefinedTypesHandler {
+    private const val cInteropPackage = "kotlinx.cinterop"
+
+    private val nativePtrClassifier = Classifier.topLevel(cInteropPackage, "NativePtr")
+
+    private val primitives = setOf(
+            KotlinTypes.boolean,
+            KotlinTypes.byte, KotlinTypes.short, KotlinTypes.int, KotlinTypes.long,
+            KotlinTypes.uByte, KotlinTypes.uShort, KotlinTypes.uInt, KotlinTypes.uLong,
+            KotlinTypes.float, KotlinTypes.double
+    )
+
+    /**
+     * kotlinx.cinterop.{primitive}Var -> kotlin.{primitive}
+     */
+    private val primitiveVarClassifierToPrimitiveType: Map<Classifier, KotlinClassifierType> =
+            primitives.associateBy {
+                val typeVar = "${it.classifier.topLevelName}Var"
+                Classifier.topLevel(cInteropPackage, typeVar)
+            }
+
+    /**
+     * @param primitiveType primitive type from kotlin package.
+     * @return kotlinx.cinterop.[primitiveType]VarOf<[primitiveType]>
+     */
+    private fun getVarOfTypeFor(primitiveType: KotlinClassifierType): ClassifierStubType {
+        val typeVarOf = "${primitiveType.classifier.topLevelName}VarOf"
+        val classifier = Classifier.topLevel(cInteropPackage, typeVarOf)
+        return ClassifierStubType(classifier, listOf(TypeArgumentStub(primitiveType.toStubIrType())))
+    }
+
+    private fun expandCOpaquePointerVar(nullable: Boolean): AbbreviationStubType {
+        val typeArgument = TypeArgumentStub(expandCOpaquePointer(nullable=false))
+        val expandedType = ClassifierStubType(KotlinTypes.cPointerVarOf, listOf(typeArgument))
```

Comment:
Isn't it required to apply nullability to `expandedType` too?
Same below.

Reply:
Also having `nullable` in `AbbreviationStubType` seems error-prone.

Reply:
Agree. I'll get rid of `AbbreviationStubType`.